### PR TITLE
Add native Qwen4-Exp support completed by Argus

### PR DESCRIPTION
**Pull request by waltstephen**
## Summary

Completed by Argus.

This PR adds native SGLang text-path support for `Qwen/Qwen3.8-Flash-Next`, whose pinned Hugging Face config declares `Qwen4ExpForConditionalGeneration` / `model_type=qwen4_exp` rather than an existing Qwen3 family architecture. The implementation registers Qwen4-Exp configs with SGLang / Transformers config loading, routes conditional generation as text-only when multimodal serving is disabled, and adds a native text model with GatedResidual, Gated DeltaNet linear attention through `RadixLinearAttention` / `GDNAttnBackend`, Qwen sparse attention masking, PLE, MoE, official-style weight-name loading, TP2 GDN cache sharding, and focused synthetic/runtime tests.

## Pinned sources

| Source | Pin / fact |
|---|---|
| Hugging Face model | `Qwen/Qwen3.8-Flash-Next` repo sha `f5d08274bafd880402bd16f5e3e6c514136ec06c` |
| Official architecture | `architectures=["Qwen4ExpForConditionalGeneration"]`, top-level `model_type="qwen4_exp"`, nested text `model_type="qwen4_exp_text"` |
| Official implementation reference | `huggingface/transformers` ref `dabae5fcb924a8eece0e727b627ca5f050b40d40`, `models/qwen4_exp` |
| License | Model card metadata: Qwen Community License 1.0; reference Transformers code: Apache-2.0 |
| SGLang base used for this branch | `f8cc1f9525c3a0bf3b14480cc76eccb79db1b4ea` |

## Correctness and serving evidence

All GPU runs below used only the assigned devices with `CUDA_VISIBLE_DEVICES=2,3`; process-local CUDA devices `0,1` were the two RTX A6000 48GB GPUs.

| Check | Evidence |
|---|---|
| Targeted native unit gate | `CUDA_VISIBLE_DEVICES=2,3 PYTHONPATH=python:test python -m pytest -q test/registered/unit/models/test_qwen4_exp_native.py` -> `16 passed, 19 warnings in 11.45s` |
| Touched-file format/style gate | `black --check` on 8 touched Python files -> `8 files would be left unchanged`; `isort --check-only` passed; `ruff check --select=F401,F821,UP037` passed on the upstream-linted touched files. `python/sglang/srt/configs/__init__.py` was included in black/isort but not the selected ruff invocation because the existing import-barrel file has unrelated pre-existing F401 behavior under a raw direct ruff call. |
| Official checkpoint weight-index coverage | Pinned `model.safetensors.index.json`: `index_weights=1658`, native params mapped `1291`, deterministic/non-text buffers skipped `367`, `unknown=0`, `missing=0`. |
| Official TP1 CUDA construction | Native official config constructed with PLE shard tensors on CPU, MoE expert tensors on CPU, LM head on `cuda:0`; peak CUDA memory `18876.996 MB`. |
| Official TP2 CUDA construction | Native official config constructed on both assigned GPUs with `attn_tp_size=2`; PLE/expert tensors on CPU, LM heads on `cuda:0` / `cuda:1`; peak CUDA memory `14899.813 MB` per rank. |
| Real TP1 native API / streaming smoke | Full checkpoint loaded through native `Qwen4ExpForConditionalGeneration`; load elapsed `829.30 s`; `/health`, `/v1/models`, native `/generate`, OpenAI `/v1/completions`, and OpenAI streaming all succeeded. Prompt `Hello from Argus` returned `output_ids=[11]`, text `","`; native e2e latency `2.2566 s`, OpenAI completion `2.39 s`, streaming `2.28 s`. |
| Real TP2 native API / streaming smoke | Full checkpoint loaded through native TP2 on the assigned GPUs; load elapsed `997.30 s`; `/health`, `/v1/models`, native `/generate`, OpenAI `/v1/completions`, and OpenAI streaming all succeeded. Same prompt returned the same `output_ids=[11]`, text `","`; native e2e latency `2.6569 s`, OpenAI completion `2.43 s`, streaming `2.43 s`. |

## Optimization evidence

These speedups are retained only for the measured scope shown here. The real-weight API smokes above are functionality/feasibility checks, not production throughput or real-weight speedup claims.

| Screen | Result |
|---|---|
| Standalone Qwen4-Exp GDN kernel-path benchmark | Baseline fp16 token-loop `127.753 ms/iter`; final Triton chunk GDN path `1.144 ms/iter`; `111.71x` speedup; `seq=512`, `batch=1`, `iters=10`, `max_abs=0.000732422`, `rel_to_ref_max=0.00116825`. |
| Synthetic integrated all-linear runtime | Full native synthetic model with `RadixLinearAttention`, `GDNAttnBackend`, `HybridReqToTokenPool`: reference loop `35.601 ms`, Radix GDN `8.511 ms`, `4.183x`. |
| Synthetic integrated mixed topology | 3 linear-attention + 1 full-attention layer pattern: reference `42.160 ms`, Radix GDN `17.753 ms`, `2.375x`. |
| Synthetic integrated fp16 aligned-conv path | Reference `36.016 ms`, Radix GDN `8.051 ms`, `4.474x` with `SGLANG_MAMBA_CONV_DTYPE=float16`. |
| Synthetic fused MoE path | Top-10 / 64-expert synthetic shape: Python MoE `21.190 ms`, fused MoE `4.087 ms`, `5.184x` MoE-path speedup; full reference loop to fused `79.809x`. |
| Synthetic TP1 after TP sharding | Reference `24.423 ms`, fused path `6.415 ms`, `3.807x`; correctness max abs `0.00390625`. |
| Synthetic TP2 after TP sharding | Both ranks reached correctness/timing; reference `25.137/25.278 ms`, fused path `6.973/6.973 ms`, `3.605x/3.625x`. |
| PLE-enabled synthetic TP1 | Reference `30.020 ms`, fused path `10.040 ms`, `2.990x`. |
| PLE-enabled synthetic TP2 | Reference `31.735/31.728 ms`, fused path `10.657/10.713 ms`, `2.978x/2.962x`. |
| PLE-enabled longer synthetic TP2 prefill | `batch=2`, `seq=128`: reference `109.348/109.319 ms`, fused path `10.863/10.844 ms`, `10.066x/10.081x`. |

## Reproduction commands

Set up an SGLang development environment with CUDA PyTorch, SGLang's normal test dependencies, and `ninja` on `PATH` for JIT kernels.

```bash
export CUDA_VISIBLE_DEVICES=2,3
export PYTHONPATH=python:test

python -m pytest -q test/registered/unit/models/test_qwen4_exp_native.py

python -m black --check \
  benchmark/qwen4_exp_runtime_benchmark.py \
  python/sglang/srt/configs/__init__.py \
  python/sglang/srt/configs/hybrid_arch.py \
  python/sglang/srt/configs/model_config.py \
  python/sglang/srt/configs/qwen4_exp.py \
  python/sglang/srt/models/qwen4_exp.py \
  python/sglang/srt/utils/hf_transformers/common.py \
  test/registered/unit/models/test_qwen4_exp_native.py

python -m isort --check-only \
  benchmark/qwen4_exp_runtime_benchmark.py \
  python/sglang/srt/configs/__init__.py \
  python/sglang/srt/configs/hybrid_arch.py \
  python/sglang/srt/configs/model_config.py \
  python/sglang/srt/configs/qwen4_exp.py \
  python/sglang/srt/models/qwen4_exp.py \
  python/sglang/srt/utils/hf_transformers/common.py \
  test/registered/unit/models/test_qwen4_exp_native.py

python -m ruff check --select=F401,F821,UP037 \
  benchmark/qwen4_exp_runtime_benchmark.py \
  python/sglang/srt/configs/hybrid_arch.py \
  python/sglang/srt/configs/model_config.py \
  python/sglang/srt/configs/qwen4_exp.py \
  python/sglang/srt/models/qwen4_exp.py \
  python/sglang/srt/utils/hf_transformers/common.py \
  test/registered/unit/models/test_qwen4_exp_native.py
```

Synthetic benchmark examples:

```bash
export CUDA_VISIBLE_DEVICES=2,3
export PYTHONPATH=python:test

python benchmark/qwen4_exp_runtime_benchmark.py \
  --enable-ple \
  --batch-size 2 \
  --seq-len 16 \
  --iters 2 \
  --warmups 1 \
  --dtype bfloat16 \
  --device cuda \
  --tp-size 1

python -m torch.distributed.run --nproc_per_node=2 --master_port=29631 \
  benchmark/qwen4_exp_runtime_benchmark.py \
  --enable-ple \
  --batch-size 2 \
  --seq-len 16 \
  --iters 2 \
  --warmups 1 \
  --dtype bfloat16 \
  --device cuda \
  --tp-size 2
```

Real-checkpoint serving smoke shape:

```bash
export CUDA_VISIBLE_DEVICES=2,3
export MODEL_PATH=/path/to/Qwen3.8-Flash-Next/snapshot

python -m sglang.launch_server \
  --model-path "$MODEL_PATH" \
  --tp 1 \
  --trust-remote-code \
  --host 127.0.0.1 \
  --port 31081

curl -s http://127.0.0.1:31081/health
curl -s http://127.0.0.1:31081/v1/models
curl -s http://127.0.0.1:31081/generate \
  -H 'Content-Type: application/json' \
  -d '{"text":"Hello from Argus","sampling_params":{"temperature":0,"max_new_tokens":1}}'
curl -s http://127.0.0.1:31081/v1/completions \
  -H 'Content-Type: application/json' \
  -d '{"model":"qwen4-exp-native","prompt":"Hello from Argus","temperature":0,"max_tokens":1}'
curl -N http://127.0.0.1:31081/v1/completions \
  -H 'Content-Type: application/json' \
  -d '{"model":"qwen4-exp-native","prompt":"Hello from Argus","temperature":0,"max_tokens":1,"stream":true}'
```

For TP2, use the same smoke shape with `--tp 2` and a different port.

## Limitations

- This PR does not implement Qwen4-Exp vision processor/tower integration. Native `Qwen4ExpForConditionalGeneration` is routed as text-only in this bounded implementation.
- The real-checkpoint TP1/TP2 evidence is API and streaming smoke coverage, not production throughput, high-concurrency serving validation, or long-context validation.
- No official Transformers deterministic logits tolerance result is claimed here. The completed correctness coverage is native deterministic synthetic coverage, official checkpoint weight-name coverage, native construction, TP1/TP2 serving smoke, and synthetic fused-path comparisons.
- Official-sized PLE and MoE tensors are intentionally CPU-resident to fit the real checkpoint on two RTX A6000 48GB GPUs. The real API smoke timings include that CPU-resident route and are not an optimized real-weight throughput claim.
- Synthetic integrated speedups use tiny random checkpoint-free models. They exercise the native SGLang model class, logits path, GDN backend, hybrid cache, TP2 cache sharding, PLE, and fused MoE where applicable, but they are not speedups on real Qwen/Qwen3.8-Flash-Next weights.
- The fp16 integrated path needs conv-state dtype alignment (`SGLANG_MAMBA_CONV_DTYPE=float16`) for the measured success route. The official config uses bf16 plus float32 Mamba/SSM dtype.
- The fused MoE path is gated to CUDA fp16/bf16, `silu`, and expert intermediate sizes divisible by 8; unsupported shapes retain the reference loop.

## Attribution

This implementation, its native SGLang integration, weight-coverage audit,
correctness checks, TP1/TP2 real-checkpoint serving smokes, optimization
benchmarks, and PR evidence were completed autonomously by Argus.

---
### CI States

Latest PR Test (Base): :x: [Run #33031211299](https://github.com/sgl-project/sglang/actions/runs/33031211299)
Latest PR Test (Extra): :x: [Run #33031211264](https://github.com/sgl-project/sglang/actions/runs/33031211264)
Latest PR Test (AMD ROCm 7.2): :x: [Run #33031211286](https://github.com/sgl-project/sglang/actions/runs/33031211286)

### CODE REVIEW

**Comment by waltstephen** [comment]
@JustinTong0323 @Qiaolin-Yu could you please review this PR and advise on the preferred merge path? This is a bounded, text-only A6000 implementation with an official-checkpoint weight-index audit (`1658` indexed tensors, `unknown=0`, `missing=0`), real TP1/TP2 API and streaming smokes, targeted tests, and reproducible GDN/MoE performance evidence. I also noticed the broader implementation in #36497; if that is the intended primary path, please let us know whether the A6000 compatibility work, weight-coverage checks, tests, or benchmark artifacts here should be folded into it. Otherwise, once required CI and CODEOWNER review are green, we would appreciate help merging this contribution. The implementation and evidence were completed by Argus.
